### ISSUE TITLE
Update go-mod-secrets to version 0.0.17 or later to include the token-refresh capability

### ISSUE THREAD

**Issue by jim-wang-intel** [edgexfoundry/docker-edgex-mongo#82]
The current mongo service is using old version of go-mod-secrets and does not have the token-refresh feature. We should update the version of go-mod-secrets to v0.0.17 to make mongo service token automatically refreshed when running in secure mode.

The current version will not refresh the Vault token automatically, hence will be expired if not using v0.0.17+ of go-mod-secrets.

**Comment by jim-wang-intel** [comment]
cc: @bnevis-i, @tingyuz, @mkbhanda, @eno-intel

**Comment by bnevis-i** [comment]
@jim-wang-intel can you clarify the effect on mongo initializes the first time, and is then restarted after the token has expired?  Does this need to be fixed to allow monogo to restart successfully?

**Comment by jim-wang-intel** [comment]
@bnevis-i if the auto-refresh is in place, under the normal circumstances, we should not see the token for mongo service expired.  It should be taken care with background auto refresh.  If the token does expired, however, one should re-generate the token either restarting the whole Edgex-stack or provides a new token file (json) I believe to allow callback func to work.

**Comment by bnevis-i** [comment]
@jim-wang-intel will wait for test results.  Expectation is that since we do not have password rotation implemented, the edgex-mongo password setup is a one-time event.   As edgex-mongo is a single-shot service (runs and exits immediately) upgrading to the new go-mod-secrets is purely hygenic.  The only requirement is that edgex-mongo-launch.sh ignore any failures here in the event the mongodb container is later restarted, since once passwords are provisioned, failure to reprovision the same password should not be grounds for failing the database startup.

**Comment by jim-wang-intel** [comment]
The edgex-mongo executable itself will not be started due to the expired token due to the line on the main.go here: https://github.com/edgexfoundry/docker-edgex-mongo/blob/master/cmd/main.go#L43 and the service itself will fail to restart since the shell script `edgex-mongo-launch.sh` exists out on the line with 
```sh
./edgex-mongo --profile=docker --confdir=res
```

Error log message:
```
level=ERROR ts=2020-02-28T23:27:07.857065344Z app=edgex-mongo source=config.go:69 msg="failed to read secret stores data for '/v1/secret/edgex/mongo/logging' path: Error found on handling secrets from underlying data-store: Received a '403' response from the secret store"
level=ERROR ts=2020-02-28T23:27:07.857103521Z app=edgex-mongo source=main.go:57 msg="edgex-mongo failed because of: Error found on handling secrets from underlying data-store: Received a '403' response from the secret store"
```
after restarting the docker edgex-mongo service once the token is expired.

So at least we need some fix on how the shell script `edgex-mongo-launch.sh` handles the token-expired error from main.go; eg. 
```sh
./edgex-mongo --profile=docker --confdir=res || true
```

So like Bryon mentioned above, confirmed that as long as the shell script ignores the error in the event the mongodb container is later restarted, upgrading to the new version of go-mod-secrets is purely hygienic.  

This is also working when the whole docker-stack restarted: `docker-compose -f docker-compose-nexus-mongo.yml up -d` . I.e. the mongo still can be restarted and running up even though the token has been expired.

Thus, this issue can be very low priority or closed.  Thanks, @bnevis-i .

**Comment by bnevis-i** [comment]
I think
```
./edgex-mongo --profile=docker --confdir=res || true
```
addresses a real issue and that we should at least fix this.

**Comment by difince** [comment]
`./edgex-mongo --profile=docker --confdir=res || true`
This line will hide/reject all the errors inside edgex-mongo that I think is not a good approach.

As I see, vault database is not deleted anymore, so edgex-mongo  go app really should be a single-shot service - it is currently not -  currently, it re-creates all the credentials on every restart.

So, I suggest to open an issue - for making it single-shot service ? 
As I see it, `edgex-mongo-launch.sh` script should have the knowledge if it needs to execute `edgex-mongo` go app at all. Thus, restarting **mongo** **database** will be needed on the very first start on the system.  
All further calls of `edgex-mongo-launch.sh` script will be responsible to just run the database with enabled authentication.

**Comment by bnevis-i** [comment]
I've been reflecting on this as well.  I don't believe the "ignore error" fix will work if we were to make a change such that edgex-mongo (or anything that  that requires access to the secret store) would wait for a valid token.  We need to be able to distinguish the bootstrapping case from the nonbootstrapping case.

Is the following a possible solution:
 - Start mongo in local-only mode
 - Check to see if users have been provisioned
 - If users have not been provisioned, run edgex-mongo to provision them
 - Restart mongo in  production mode

The above should handle the case if mongo service fails for some reason (including a developer stopping and restarting the service) and is restarted (perhaps automatically).

I do not know if we need at this point to support an upgrade flow, where we later add services and users to an existing database.
-

**Comment by jim-wang-intel** [comment]
> 
> 
> `./edgex-mongo --profile=docker --confdir=res || true`
> This line will hide/reject all the errors inside edgex-mongo that I think is not a good approach.
> 

I agree with you on this.  For the short-term, maybe we should give an different or specific error code for this case.  With that specific error code, then maybe in the script we could use that to determine what to do next.

> As I see, vault database is not deleted anymore, so edgex-mongo go app really should be a single-shot service - it is currently not - currently, it re-creates all the credentials on every restart.
> 
> So, lets open an issue for making it single-shot service ?

Should it be?  IMHO, from user's mental model perspective, isn't he/she expecting to see mongo service to be up and running?

> As I see it, [edgex-mongo-launch.sh](https://github.com/edgexfoundry/docker-edgex-mongo/blob/master/bin/edgex-mongo-launch.sh) script should have the knowledge if it needs to execute `edgex-mongo` go app at all.
> Thus, executing `edgex-mongo`, as well restarting mongo database itself, both will be needed on the very first start of the system.
> All further calls of `edgex-mongo-launch.sh` script will be responsible to just run the database with enabled authentication.

Agree.

**Comment by bnevis-i** [comment]
> So, lets open an issue for making it single-shot service ?

Sure go ahead and close this one and re-open a new one with your proposal in it and I'll do my best to try to identify all the corner cases.

**Comment by jim-wang-intel** [comment]
Close this issue as per requirement of desirable non-renewed token for mongo itself.

**Comment by bnevis-i** [comment]
What the new issue #?

**Comment by difince** [comment]
> Is the following flow a possible solution?

> Start mongo in local-only mode
Check to see if users have been provisioned
If users have not been provisioned, run edgex-mongo to provision them
Restart mongo in production mode

What I was thinking for now is .. once edgex-mongo finish its work at the end of its execution some hidden file to be created, so the existence of this hidden file could be used in the `edgex-mongo-launch.sh` for distinguishing  bootstrapping from the nonbootstrapping case.


> I do not know if we need at this point to support an upgrade flow, where we later add services and users to an existing database.

For supporting upgrade functionality, the proposal above will not work. I guess we will need to add the version field into the database itself as well. And in the begging of edgex-mongo execution to  validate if there is a change in the versioning. 
I would also think we could consider not only the software version ( as `v0.0.17` is for go-mod-secret), but in addition to introduce `db_version`. Because not each software version change leads to database changes. But I also do not know if this is the correct time to think of it.  

> Should it be? IMHO, from user's mental model perspective, isn't he/she expecting to see mongo service to be up and running

I am not sure if I got your point. But even now, after `edgex-mongo` execution finish, the container `edgex-mongo` stays up and running. 

I see you have closed this issue, but isn't it good to use the latest-greatest version of go-mod-secret ? I was thinking to update go-mod-secret version and in addition eventually to open another issue ... 
At first glance I saw that changing the version of go-mod-secret in `go.mod` file won't be enough... and some other code change need to be done as well.

**Comment by bnevis-i** [comment]
> I see you have closed this issue, but isn't it good to use the latest-greatest version of go-mod-secret ?

The change was proposed ostensibly to get the new version with the token refresh logic.  Unless edgex-mongo keeps running the the background to keep the token fresh, it won't matter.  That's why Jim closed the issue.

> What I was thinking for now is .. once edgex-mongo finish its work at the end of its execution some hidden file to be created, so the existence of this hidden file could be used in the edgex-mongo-launch.sh for distinguishing bootstrapping from the nonbootstrapping case.

Works for me.

**Comment by difince** [comment]
aa, you mean if the service edgx-mongo does not exist, but stays up and running - the token refresh logic is going to work and there won't be a problem  ?

**Comment by bnevis-i** [comment]
> aa, you mean if the service edgx-mongo does not **exit**, but stays up and running - the token refresh logic is going to work and there won't be a problem ?

Yes, that will prevent the failure as well.

**Comment by difince** [comment]
but it does not make sense to keep a service running if it is doing anything ... ?

**Comment by bnevis-i** [comment]
> but it does not make sense to keep a service running if it is doing anything ... ?

Other than keeping the token fresh so next time the container restarts it can reprovision the passwords it already provisioned last time.

**Comment by jim-wang-intel** [comment]
I closed this is because it is desirable not to have the mongo's Vault token refreshed as I go that impression from @bnevis-i.  

As far as the new issue is concerned, I'll let Diana @difince create one as the original thought is from her.  

Thanks!

**Comment by bnevis-i** [comment]
> I closed this is because it is desirable not to have the mongo's Vault token refreshed as I go that impression from @bnevis-i.

Yes, that too.  No reason to keep the token around once provisioning is done we shan't need it any more.

**Comment by difince** [comment]
Ok, but I guess that some day we will get back to this issue ..
